apps/emqx/src/emqx_schema.erl

```diff
@@ -2376,3 +2356,224 @@ non_empty_string(<<>>) -> {error, empty_string_not_allowed};
 non_empty_string("") -> {error, empty_string_not_allowed};
 non_empty_string(S) when is_binary(S); is_list(S) -> ok;
 non_empty_string(_) -> {error, invalid_string}.
+
+%% @doc Make schema for 'server' or 'servers' field.
+%% for each filed, there are three passes:
+%% 1. converter: Normalize the value.
+%%               This normalized value is stored in MEQX's raw config.
+%% 2. validator: Validate the normalized value.
+%%               Besides checkin if the value can be empty or undefined
+%%               it also calls the 3rd pass to see if the provided
+%%               hosts can be successfully parsed.
+%% 3. parsing: Done at runtime in each module which uses this config
+servers_sc(Meta0, ParseOpts) ->
+    Required = maps:get(required, Meta0, true),
+    Meta = #{
+        required => Required,
+        converter => fun convert_servers/2,
+        validator => servers_validator(ParseOpts, Required)
+    },
+    sc(string(), maps:merge(Meta, Meta0)).
+
+%% @hidden Convert a deep map to host:port pairs.
+%% This is due to the fact that a host:port string
+%% often can be parsed as a HOCON struct.
+%% e.g. when a string from environment variable is `host.domain.name:80'
+%% without escaped quotes, it's parsed as
+%% `#{<<"host">> => #{<<"domain">> => #{<<"name">> => 80}}}'
+%% and when it is a comma-separated list of host:port pairs
+%% like `h1.foo:80, h2.bar:81' then it is parsed as
+%% `#{<<"h1">> => #{<<"foo">> => 80}, <<"h2">> => #{<<"bar">> => 81}}'
+%% This function is to format the map back to host:port (pairs)
+%% This function also tries to remove spaces around commas in comma-separated,
+%% `host:prot' list, and format string array to comma-separated.
+convert_servers(HoconValue, _HoconOpts) ->
+    convert_servers(HoconValue).
+
+convert_servers(undefined) ->
+    %% should not format 'undefined' as string
+    %% not to throw exception either
+    %% (leave it to the 'required => true | false' check)
+    undefined;
+convert_servers(Map) when is_map(Map) ->
+    try
+        List = convert_hocon_map_host_port(Map),
+        string:join(List, ",")
+    catch
+        _:_ ->
+            throw("bad_host_port")
+    end;
+convert_servers([H | _] = Array) when is_binary(H) orelse is_list(H) ->
+    %% if the old config was a string array
+    %% we want to make sure it's converted to a comma-separated
+    iolist_to_binary([[I, ","] || I <- Array]);
```

Comment:
Q: the `is_map` clause returns an Erlang string (list), but the list of binaries/strings will return a binary.
Is either type fine in this case?

Also, the binary/string list clause also will introduce a trailing comma, it seems:

```erlang
3> iolist_to_binary([[I, ","] || I <- ["a", "b"]]). 
<<"a,b,">>
```

Reply:
yes, string and binary both works.
but I have changed to always return binary to avoid confusion.